dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/utils/MemoizationUtil.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2022-2023 Google LLC
+ * Copyright 2013-2021 CompilerWorks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.edwmigration.dumper.application.dumper.utils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import org.apache.commons.io.function.IOFunction;
+
+/** Utility class for memoization. */
+public class MemoizationUtil {
+
+  private static class OptionalMemoizer<T, K, R> implements IOFunction<T, Optional<R>> {
```

Comment:
IIUC the `OptionalMemoizer` class is needed, so that retries are possible. It seems that these two concerns (caching and retrying) are mixed in this solution. I would suggest separating them by using a `LoadingCache` from guava, so that retrying could be performed independently (internally in the loading function).

https://github.com/google/guava/wiki/CachesExplained

Reply:
No, retries are handled by the Apache Commons Http Client. If all retries fail, an IOException is raised.

This is just to say that the delegate should be checked again if the optional is empty. It is not strictly necessary here as the information does not change on the delegate but there is also no expensive IO happening.

Reply:
The PR description is: "The cloud extractor can do multiple dumps and will likely require the same driver for all runs (e.g. metadata and logs for the same DB). Thus, the driver for it only needs to be downloaded once."

If the download is unsuccessful (IOException happens), then the process will stop.

If the download is successful, then the driver path will be saved in the map and there won't be a need to call the delegate the second time.

> (...) delegate should be checked again if the optional is empty.

Could you point to the code, where the second check would happen (or if the code is not ready yet, then describe such a use-case)?

Reply:
I don't understand. As I said, the Optional is not necessary here and hence I removed it. What are we debating now?

Reply:
Are you wondering about why the retriever might be called a 2nd time? That is because there can be multiple connectors in the cloud extractor configuration. See https://github.com/google/dwh-migration-tools/blob/main/dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/clouddumper/Main.java#L79

The use case is that you want to download metadata and query logs in one go.

Reply:
2023-07-28 10:07 GMT you wrote:
> This is just to say that the delegate should be checked again if the optional is empty.

2023-07-28 10:26 GMT I wrote:
> Could you point to the code, where the second check would happen (or if the code is not ready yet, then describe such a use-case)?

2023-07-28 11:38 GMT you force-pushed the removal of Optional from Memoizer:
https://github.com/google/dwh-migration-tools/compare/9246c978283a982341ef5622cc60f9d9b6991a4d..c64143fc325b7bc7f3e427b103678bd449971d38

2023-07-29 10:52 GMT you wrote:
I don't understand. As I said, the Optional is not necessary here and hence I removed it.

---

I just wanted to point out that IIUC the delegate will never be called a second time for the same driver. You removed the Optional from the Memoizer after my comment, so the case you described as "the delegate should be checked again if the optional is empty" won't happen, so it's all good now.